packages/devtools/lib/src/timeline/timeline_protocol.dart

```diff
@@ -322,22 +397,120 @@ class TimelineEvent {
 
   bool get isCpuEvent => type == TimelineEventType.cpu;
   bool get isGpuEvent => type == TimelineEventType.gpu;
 
+  bool get isPipelineProduceFlow => _hasChild('Engine::BeginFrame');
+  bool get isPipelineConsumeFlow => _hasChild('PipelineConsume');
+
+  /// Whether there is a child with the given name [childName] is contained
+  /// somewhere in the subtree [children].
+  bool _hasChild(String childName) {
+    bool childFound = false;
+
+    void findChild(TimelineEvent event, String childName) {
+      if (event.name.contains(childName)) {
+        childFound = true;
+      }
+      if (childFound) return;
+      for (TimelineEvent e in event.children) {
+        findChild(e, childName);
+      }
+    }
+
+    findChild(this, childName);
+    return childFound;
+  }
+
+  /// Adds a child event.
+  ///
+  /// If we know the child's location (knownLocation == true)
+  /// add the child directly via [_addChild]. If we do not know the child's
+  /// location, find it's proper place in the subtree and add the child there.
+  void addChild(TimelineEvent child, {bool findChildLocation = false}) {
+    bool newChildAdded = false;
+
+    // Places the child in it's correct position amongst the other children.
+    void _putChildInSubtree(TimelineEvent root, TimelineEvent newChild) {
+      // [root] is a leaf. Add child here.
+      if (root.children.isEmpty) {
+        root._addChild(child);
+        newChildAdded = true;
+      }
+
+      if (newChildAdded) return;
+
+      final List<TimelineEvent> _children =
+          List<TimelineEvent>.from(root.children);
+      for (int i = 0; i < _children.length; i++) {
+        final TimelineEvent currentChild = _children[i];
+
+        // [newChild] is the parent of [currentChild].
+        if (newChild._isParentOf(currentChild)) {
+          // Link [currentChild] with its correct parent [newChild].
+          newChild._addChild(currentChild);
+
+          // Unlink [child] from its incorrect parent [this].
+          root.children.remove(currentChild);
+
+          // Link [newChild] with its correct parent [this].
+          if (!newChildAdded) _addChild(newChild);
+
+          newChildAdded = true;
+        }
+
+        // [child] is the parent of [newChild].
+        if (currentChild._isParentOf(newChild)) {
+          // Recurse on [currentChild]'s subtree.
+          _putChildInSubtree(currentChild, newChild);
+        }
+      }
+
+      // If we have not added the child at this point, [currentChild] and
+      // [newChild] are siblings.
+      if (!newChildAdded) {
+        root._addChild(newChild);
+        newChildAdded = true;
+        return;
+      }
+    }
+
+    if (findChildLocation) {
+      _putChildInSubtree(this, child);
+    } else {
+      _addChild(child);
+    }
+  }
+
+  void _addChild(TimelineEvent child) {
+    children.add(child);
+    child.parent = this;
+  }
+
+  // TODO(kenzie): consider comparing with an epsilon for endTime.
```

Comment:
we need to use the epsilons everywhere or nowhere. I'm interested in understanding why the epsilon is currently needed.

Reply:
For some reason, events can be slightly off of the frame start and end times by several microseconds. Happy to decrease the epsilon at some point, but I made it larger than I thought we'd need to be safe. It seems to generally be off by just a few microseconds, but I did see a couple cases where it was off by ~20-30. 

Could be an issue with when the frame start/end events are triggered, I'm not sure. But the behavior of being slightly out of bounds is consistent.

Reply:
I'm going to leave this TODO here because I'm not sure if parent/child relationships will extend boundaries like the cpu/gpu flows extend the frame start/end times. If they do, it will likely be a much smaller epsilon than what we are using for frame start/end times.

Reply:
sg